src/AudioAnalysisTools/DSP/PcaWhitening.cs

```diff
@@ -0,0 +1,212 @@
+            //To do so, we need eigenVectors, and the number of columns of the projected data
+            double[,] projectionMatrix = GetProjectionMatrix(eigenVectors, columns);
+
+            //write the projection matrix to disk
+            /*
+            //FIRST STEP: sort the eigenvectors based on the eigenvalue
+            var eigPairs = new List<Tuple<double, double[]>>();
+
+            for (int i = 0; i < eigneValues.GetLength(0); i++)
+            {
+                eigPairs.Add(Tuple.Create(Math.Abs(eigneValues[i]), GetColumn(eigenVectors, i)));
+            }
+
+            //sort in descending order based on the eigenvalues
+            eigPairs.Sort((x, y) => y.Item1.CompareTo(x.Item1));
+            */
+
+            return new Tuple<double[,], double[,], double[,], int>(projectionMatrix, reversion, eigenVectors, components);
+        }
+
+        // RMS Normalization
+        public static double[,] RmsNormalization(double[,] matrix)
+        {
+            double s = 0;
+            double[,] normSpec = new double[matrix.GetLength(0), matrix.GetLength(1)];
+            for (int i = 0; i < matrix.GetLength(0); i++)
+            {
+                for (int j = 0; j < matrix.GetLength(1); j++)
+                {
+                    s += matrix[i, j] * matrix[i, j];
+                }
+            }
+
+            double rms = Math.Sqrt(s / (matrix.GetLength(0) * matrix.GetLength(1)));
+
+            for (int i = 0; i < matrix.GetLength(0); i++)
+            {
+                for (int j = 0; j < matrix.GetLength(1); j++)
+                {
+                    normSpec[i, j] = matrix[i, j] / rms;
+                }
+            }
+
+            return normSpec;
+        }
+
+        //retrieving a full column of a matrix
+        //colNo is the column index we want to access
+        public static double[] GetColumn(double[,] matrix, int colNo)
+        {
+            double[] col = new double[matrix.GetLength(0)];
+            for (int r = 0; r < matrix.GetLength(0); r++)
+            {
+                col[r] = matrix[r, colNo];
+            }
+
+            return col;
+        }
+
+        //retrieving a full row of a matrix
+        //rowNo is the column index we want to access
+        public static double[] GetRow(double[,] matrix, int rowNo)
+        {
+            double[] row = new double[matrix.GetLength(1)];
+            for (int c = 0; c < matrix.GetLength(1); c++)
+            {
+                row[c] = matrix[rowNo, c];
+            }
+
+            return row;
+        }
+
+        //Build the Projection Matrix
+        //To do so, we need eigenVectors
+        //and the number of columns of the projected data which is the number of outputs (principle components) used to transform the data
+        public static double[,] GetProjectionMatrix(double[,] eigenVector, int numberOfOuputs)
+        {
+            double[,] projMatrix = new double[eigenVector.GetLength(0), eigenVector.GetLength(1)];
```

Comment:
I see this type of construct a lot in your code. 


Add this method to `DoubleSquareArrayExtensions`:
```
public static double[] EmptyCopy(this double[,] matrix) {
   return new double[matrix.Rows(), matrix.coumns()];
}
```

Use like this `eigenVector.EmptyCopy()`.